Library/src/Systems/Users/UserSystem.cpp

```diff
@@ -142,22 +144,35 @@ void UserSystem::Login(const csp::common::String& UserName,
 			{
 				NotifyRefreshTokenHasChanged();
 
+				std::promise<csp::multiplayer::ErrorCode> ConnectPromise;
+				std::future<csp::multiplayer::ErrorCode> ConnectFuture = ConnectPromise.get_future();
+
 				csp::multiplayer::MultiplayerConnection::ErrorCodeCallbackHandler ErrorCallback
-					= [Callback, LoginStateRes](csp::multiplayer::ErrorCode ErrCode)
+					= [Callback, LoginStateRes, &ConnectPromise](csp::multiplayer::ErrorCode ErrCode)
 				{
-					if (ErrCode != csp::multiplayer::ErrorCode::None)
-					{
-						CSP_LOG_ERROR_FORMAT("Error connecting MultiplayerConnection: %s", csp::multiplayer::ErrorCodeToString(ErrCode).c_str());
-
-						Callback(LoginStateRes);
-						return;
-					}
-
-					Callback(LoginStateRes);
+					ConnectPromise.set_value(ErrCode);
 				};
 
 				auto* MultiplayerConnection = SystemsManager::Get().GetMultiplayerConnection();
 				MultiplayerConnection->Connect(ErrorCallback);
+
+				// Ensure this doesn't block to prevent wasm errors
+				std::async(std::launch::deferred,
+						   [&ConnectFuture, Callback, LoginStateRes]
```

Comment:
[Comment] My big technical worry with this is twofold.

1. With `std::async` launch modes, I recall when they were first coming out talk about whether or not they were actually executed on a different thread being implementation defined. 
2. std::launch::deferred may still block right? It's lazily evaluated, dosen't it just execute immediately on the calling thread, since we don't attempt to defer requesting its result via a future? I'm actually confused by this, my reading of this code is that the async block does nothing. What am I missing?

Reply:
My big non-technical worry is, as we've discussed, that this is a very specific solve for the failing multiplayer connection, and I'm not sure it will scale to all callbacks that need to be executed off-thread. I'm willing to grant it might though. The thing I want to avoid is multiple patterns emerging in the complicated threading space, especially when it's non-obvious from a purely c++ perspective why it's being done this way at all!

Reply:
My big _other_ worry not specific to this implementation is we can't put automated regression tests in for this ... and like ... this is such a prime candidate for needing tests like that, due to the complexity of the space and the low level of observability. 🫤 😖

Reply:
The usage of async here is to prevent the callback from firing in that code path. Something that the client is doing in that callback and that point is causing the error about blocking the main thread. By deferring, we are preventing blocking at that point in time.

Reply:
I'm wondering _how_ though. According to the [docs ](https://en.cppreference.com/w/cpp/thread/launch)for `std::launch::deferred`, 
> the task is executed on the calling thread the first time its result is requested (lazy evaluation)

Normally, I would expect this to be controlled by the future returned from std::async, something like
```cpp
auto future = std:async(std::launch::deferred, callable);
... some intermediate steps
result = future.get(); //The callback is called here.
```

I don't understand what the expectation is here when we just launch a `std::async` as deferred without any controlling future? When does the async block get executed?